design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+```
+  static Client create(ClientOptions clientOptions) {
+    return new ClientImpl(clientOptions);
+  }
+
+  static Client create(ClientOptions clientOptions, Vertx vertx) {
+    return new ClientImpl(clientOptions, vertx);
+  }
+```
+
+The Java client will be implemented as a Vert.x HttpClient. We expose a constructor that allows users to provide their own `Vertx` instance
+in order to take advantage of a shared connection pool and other properties if desired.
+
+`ClientOptions` will initially be as follows:
+```
+public interface ClientOptions {
+
+  ClientOptions setHost(String host);
+
+  ClientOptions setPort(int port);
+
+  ClientOptions setUseTls(boolean useTls);
+
+  ClientOptions setUseClientAuth(boolean useClientAuth);
+
+  ClientOptions setTrustStore(String trustStorePath);
+
+  ClientOptions setTrustStorePassword(String trustStorePassword);
+
+  ClientOptions setKeyStore(String keyStorePath);
+
+  ClientOptions setKeyStorePassword(String keyStorePassword);
+
+  ClientOptions setBasicAuthCredentials(String username, String password);
+
+  String getHost();
+
+  int getPort();
+
+  boolean isUseTls();
+
+  boolean isUseClientAuth();
+
+  boolean isUseBasicAuth();
+
+  String getTrustStore();
+
+  String getTrustStorePassword();
+
+  String getKeyStore();
+
+  String getKeyStorePassword();
+
+  String getBasicAuthUsername();
+
+  String getBasicAuthPassword();
+
+  ClientOptions copy();
+
+  static ClientOptions create() {
+    return new ClientOptionsImpl();
+  }
+}
+```
+
+We can always add additional configuration options later. We may also wish to expose the Vert.x `HttpClientOptions` for
+advanced users that wish to provide custom configs.
+
+### Transient queries -- Streaming
+
+The `Client` interface will provide the following methods for streaming the results of a transient (push or pull) query:
+```
+  /**
+   * Execute a query (push or pull) and receive the results one row at a time.
+   *
+   * @param sql statement of query to execute.
+   * @return query result.
+   */
+  CompletableFuture<QueryResult> streamQuery(String sql);
+
+  /**
+   * Execute a query (push or pull) and receive the results one row at a time.
+   *
+   * @param sql statement of query to execute.
+   * @param properties query properties.
+   * @return query result.
+   */
+  CompletableFuture<QueryResult> streamQuery(String sql, Map<String, Object> properties);
+```
+where `QueryResult` is as follows:
+```
+/**
+ * The result of a query (push or pull), streamed one row at time. Records may be consumed by either
+ * subscribing to the publisher or polling (blocking) for one record at a time. These two methods of
+ * consumption are mutually exclusive; only one method may be used (per QueryResult).
+ */
+public interface QueryResult extends Publisher<Row> {
+
+  List<String> columnNames();
+
+  List<String> columnTypes();
+
+  String queryID();
+
+  /**
+   * Block until a row becomes available.
+   *
+   * @return the row.
+   */
+  Row poll();
```

Comment:
For pull queries, this would block forever if the end of the result is reached? That might be a bad user experience. We could of course throw an exception, but that seems equally bad?

How does the JDBC interface handle this case (the also return a finite `ResetSet` one can iterate through)

Reply:
I believe it will return when the subscription is completed.

Reply:
What subscription? If you call `poll()` you don't subscribe?

Reply:
If you call `poll()` there's an internal subscriber. Regardless of implementation, we should make it so that when the publisher is closed, `poll()` returns (which is the essence of what Tim is saying).

Reply:
So people need to add `null` check on the result of `poll()` to cover this case?

Reply:
They would have to do that anyway. null is always a valid return value of poll.

Reply:
Can you elaborate? Why that?